tsdb/engine/tsm1/tombstone.go

```diff
@@ -79,34 +87,56 @@ func (t *Tombstoner) Delete() error {
 	if err := os.RemoveAll(t.tombstonePath()); err != nil {
 		return err
 	}
+	t.statsLoaded = false
 	return nil
 }
 
 // HasTombstones return true if there are any tombstone entries recorded.
 func (t *Tombstoner) HasTombstones() bool {
-	stat, err := os.Stat(t.tombstonePath())
-	if err != nil {
-		return false
-	}
-
-	return stat.Size() > 0
+	t.mu.RLock()
+	b := len(t.fileStats) > 0 && t.fileStats[0].Size > 0
+	t.mu.RUnlock()
+	return b
 }
 
 // TombstoneFiles returns any tombstone files associated with this TSM file.
 func (t *Tombstoner) TombstoneFiles() []FileStat {
+	t.mu.RLock()
+	if t.statsLoaded {
+		stats := t.fileStats
+		t.mu.RUnlock()
+		return stats
+	}
+	t.mu.RUnlock()
+
 	stat, err := os.Stat(t.tombstonePath())
-	if err != nil {
+	if os.IsNotExist(err) {
```

Comment:
I think this needs to be `t.stasLoaded = os.IsNotExist(err)` since error will never be nil.